### DISCUSSION TITLE
SH-RPi NMEA 2000 interface restarting

### DISCUSSION THREAD

**Discussion by mairas** [hatlabs/discussions#17] [General]
Hi all,

**TL;DR: Apply the modification described below to improve your NMEA 2000 interface reliability.**

The default Linux CAN interface configuration (at least for the chip used on SH-RPi) makes the interface enter a "bus-off" error state if, for example, the SH-RPi NMEA 2000 interface is disconnected momentarily. This can be seen in the `dmesg` command output, but the immediate result is that data stops flowing until the device is rebooted.

I finally stumbled upon the [Linux SocketCAN documentation](https://www.kernel.org/doc/html/latest/networking/can.html#starting-and-stopping-the-can-network-device) and found the correct way to fix this problem. The SocketCAN interface can be set to automatically restart after a short time with the following command:

```
ip link set can0 type can restart-ms 100
```

To apply this command permanently, you should modify the file `/etc/network/interfaces.d/can0` contents as follows:

```
auto can0
iface can0 can static
        bitrate 250000
        pre-up ip link set can0 type can restart-ms 100
        up /sbin/ifconfig can0 txqueuelen 100
```

I have added the line in the automated installation script, so for anyone doing a fresh install, the change is already included.
